### PR TITLE
MGS: Rework serialization of large-data-bearing packets

### PR DESCRIPTION
**Pull request by jgallagher**
Prior to this PR, we had two variants of MGS <-> SP messages that had large, fixed-size arrays in them (`SerialConsole` for serial console data and `UpdateChunk` for a part of an update image). This had a few undesirable properties:

1. Memory usage on the SP was higher than it needed to be due to needing to instantiate an instance of these types and then copy the data into the arrays during deserialization; it should've been able to refer to the data as a borrowed slice from the receive buffer. hubpack is specifically designed with this in mind.
2. We had to suppress clippy's large enum variant warning for similar reasons.
3. Serializing these packets had some clever machinery to keep from sending zero-filled, full length arrays in every UDP message, but (a) clever serde code is not the most readable and (b) network packet size is not as much of a concern as SP memory usage.

This PR changes the messaging format to allow for packets to be immediately followed by (length-prefixed) binary data, which allows the SP to deserialize only the packet header and then _borrow_ the binary data. This is most obvious in the change to the `SpHandler` trait. Before this PR, the `serial_console_write()` and `update_chunk()` methods took `SerialConsole`/`UpdateChunk`s by value which contained large arrays; now they receive arguments for the metadata and a separate argument for the data (a `&[u8]` which is borrowed from the UDP receive buffer).

In the process of testing these changes I ran into a nasty async bug in MGS's serial console forwarding; it's fixed as part of this PR in a6b96598762b8a1fc549199fd5c1efd6e590fe3d.

### CODE REVIEW

File: gateway-sp-comms/src/single_sp.rs [hunk 1]
```diff
@@ -836,6 +958,24 @@ fn sp_busy_policy() -> backoff::ExponentialBackoff {
     }
 }
 
+// Helper trait to provide methods on `io::Cursor` that are currently unstable.
+trait CursorUnstable {
```
Comment (andrewjstone) [inline gateway-sp-comms/src/single_sp.rs:962]: Nit: When I was reading the code above, I thought that `stable` was some type of property of the cursor itself. Maybe just rename this to `CursorHelper` or something so as not to give that impression.

File: gateway/faux-mgs/src/usart.rs [hunk 2]
```diff
@@ -39,14 +38,14 @@ pub(crate) async fn run(
     let mut out_buf = StdinOutBuf::new(raw);
     let mut flush_delay = FlushDelay::new(stdin_buffer_time);
     let console = sp
-        .serial_console_attach()
+        .serial_console_attach(SpComponent::SP3)
```
Comment (andrewjstone) [inline gateway/faux-mgs/src/usart.rs:41]: Looking at this, I think `SpComponent::SP3_HOST_CPU` would be **much** clearer.

**Review by andrewjstone — APPROVED** [review]
This is quite an improvement! Love the net-negative line count. 

Mostly just nits around naming from me.

File: gateway-messages/src/lib.rs [hunk 3]
```diff
@@ -499,159 +349,51 @@ impl TryFrom<&str> for SpComponent {
...
-        assert_eq!(&deserialized.data[..line.len()], line);
-    }
-
-    #[test]
-    fn serial_console_data_length_fits_in_u16() {
-        // this is just a sanity check that if we bump `MAX_DATA_PER_PACKET`
-        // above 65535 we also need to change the type of `SerialConsole::len`
-        assert!(SerialConsole::MAX_DATA_PER_PACKET <= usize::from(u16::MAX));
-    }
+/// Sealed trait restricted the types that can be passed to
```
[the remark this answers is not in the document]
Comment (jgallagher) [reply gateway-messages/src/lib.rs:352]: `restricting` I think is what I meant :joy:

File: gateway-sp-comms/src/single_sp.rs (hunk 1 above)
Comment (jgallagher) [reply gateway-sp-comms/src/single_sp.rs:962]: I'll go with `CursorExt` since it's sort of like an extension trait.

**Pull request merged**